### PR TITLE
Avoid thrown exception when building generic classes

### PR DESCRIPTION
**Pull request by eladlaufer**
Example:

```java
public interface InstanceBuilder<Member> extends Builder<Instance<Member>> {
      ...
}
```
throws:
```
java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to java.lang.Class
	at com.fluentinterface.utils.GenericsUtils.getDeclaredGenericType(GenericsUtils.java:21)
	at com.fluentinterface.ReflectionBuilder$InternalBuilderAsSuperClassDelegate.getClassBuiltBy(ReflectionBuilder.java:111)
	at com.fluentinterface.ReflectionBuilder.getBuiltClass(ReflectionBuilder.java:69)
	at com.fluentinterface.ReflectionBuilder.create(ReflectionBuilder.java:85)
```

### CODE REVIEW

**Comment by davidmarquis** [comment]
I fail to see why we'd need a typed interface as the builder interface. Can you explain for what use case this is for?

In your test, you a builder with the following signature?

``` java
public interface PersonBuilder extends Builder<Person> {
    PersonBuilder withId(String id);
}
```

or to your example in the pull request description:

``` java
public interface MemberBuilder extends Builder<Member> {
      ...
}
```

**Comment by eladlaufer** [comment]
In the test example, if you add:

```java
        TypedID<String> example = personTypedIDBuilder
                .withId("John Smith")
                .build();
```

The complier will throw an error, in this example I use generics to make sure that I don't assign an ID of one type to another type by mistake - It is safer then using a string directly.

Of course the example is very simple, but I do have some very complex objects the follows this pattern - And I really don't want to fallback to manually creating builders for each :)

**Comment by davidmarquis** [comment]
The compiler emits an error if you change `TypedID<Person>` to `TypedID<String>` because the `personTypedIDBuilder` is declared as being of type `TypedID<Person>` just above.

If you change the snippet to the following, it compiles properly:

```
    private TypedIDBuilder<String> personTypedIDBuilder = implementationFor(TypedIDBuilder.class)
            .usingFieldsDirectly()
            .create();

    @Test
    public void shouldSetPropertyValueString() {

        TypedID<String> built = personTypedIDBuilder
                .withId("John Smith")
                .build();

        assertThat(built.getId(), is("John Smith"));
    }
```

in fact, if you change the type of TypeIDBuilder to anything else, your test will still pass. Using `Integer` as the type still passes the test:

``` java
    private TypedIDBuilder<Integer> personTypedIDBuilder = implementationFor(TypedIDBuilder.class)
            .usingFieldsDirectly()
            .create();

    @Test
    public void shouldSetPropertyValueString() {

        TypedID<Integer> built = personTypedIDBuilder
                .withId("John Smith")
                .build();

        assertThat(built.getId(), is("John Smith"));
    }
```

But this is because the `withId` method on `TypeIDBuilder` should instead be:

``` java
TypedIDBuilder<T> withId(T id);
```

rather than

``` java
TypedIDBuilder<T> withId(String id);
```

The static nature of Java would require that you create a concrete builder for each object you want a builder for... Usually you'll have different attributes for each object anyways so at some point you need a concrete builder interface for each specific object. I'm probably missing something...

If possible, can you provide a concrete example of the pattern you're using?

**Comment by eladlaufer** [comment]
First, I'd like to thank you for your patient :)

I have a class:

```java
public class ServiceInstance<T>
{
    private StringID<ProvidedService<T>> serviceName;
    private StringID<ServiceInstance<T>> id;
    private GUID<T> ownerId;
    private Attributes attributes;

    ... omitted getters ...

    public interface ServiceInstanceBuilder<T> extends Builder<ServiceInstance<T>> {
        ServiceInstanceBuilder<T> withServiceName(StringID<ProvidedService<T>> serviceName);
        ServiceInstanceBuilder<T> withId(StringID<ServiceInstance<T>> id);
        ServiceInstanceBuilder<T> withOwnerId(GUID<T> ownerId);
        ServiceInstanceBuilder<T> withAttributes(Attributes attributes);
    }

    public static ServiceInstanceBuilder<Member> newMemberServiceInstanceBuilder() {
        //noinspection unchecked
        return ReflectionBuilder.implementationFor(ServiceInstanceBuilder.class).usingFieldsDirectly().create();
    }
}
```

As you can see not only the encapsulating class is typed - the type is used to make sure that all the fields are proper as well.

Currently the following code fails at runtime with the above exception:

```java
public Provision<Member> createOrUpdateProvision(ProvisionRequest<Member> provisionRequest) {

        ServiceInstance<Member> serviceInstance = newMemberServiceInstanceBuilder()
                .withServiceName(provisionRequest.getServiceName())
                .withId(provisionRequest.getInstanceId())
                .withOwnerId(provisionRequest.getOwnerId())
                .withId(provisionRequest.getInstanceId())
                .withAttributes(provisionRequest.getAttributes())
                .build();

        ... omitted rest ...

```

**Comment by davidmarquis** [comment]
OK I think I understand what you're trying to achieve. 

Now my questioning is more around whether we can assume that a typed interface as a builder's type *always* means we want to build the typed interface's type (ughhh)... I'm not sure we can make this assumption as part of this library.

Take this example (which closely mimic your use case but with non-specifics: 

``` java
public interface SomeGenericInterfaceBuilder<T> extends Builder<SomeGenericInterface<T>> {
    ...
}
```

What kind of instance should this library (as a generic, re-usable tool) be building for the example above? An instance of `T` or an instance of `SomeGenericInterface`? I don't think we can provide any guarantee that a developer would always want to build an instance of `T` when designing their interface like the above.

**Comment by eladlaufer** [comment]
If I understand the question correctly then the above should define a builder that its ```build()``` method returns an instance of ```SomeGenericInterface<T>```... T being defined by the developer when he asks for a new builder.

I hope this makes sense :)

**Comment by davidmarquis** [comment]
Gotcha! Sorry it took a while... then your change makes sense. Can you please cleanup the test code a bit (those auto-generated comments are nasty) and comment/refactor to make the ParameterizedType check's intent clearer? Thanks!

**Comment by eladlaufer** [comment]
Will do :)

Thanks.

File: src/test/java/com/fluentinterface/domain/TypedID.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.fluentinterface.domain;
+
+/**
+ * YouTab Media 2011 Ltd.
+ * Created by eladlaufer on 26/10/2015.
+ */
+public class TypedID<T> {
+
+    private String id;
```
Comment (davidmarquis) [inline src/test/java/com/fluentinterface/domain/TypedID.java]: Should that be typed as `T`?

File: src/test/java/com/fluentinterface/domain/TypedIDBuilder.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.fluentinterface.builder.Builder;
+
+/**
+ * YouTab Media 2011 Ltd.
+ * Created by eladlaufer on 26/10/2015.
+ */
+public interface TypedIDBuilder<T> extends Builder<TypedID<T>> {
+
+    TypedIDBuilder<T> withId(String id);
```
Comment (davidmarquis) [inline src/test/java/com/fluentinterface/domain/TypedIDBuilder.java]: `...(T id);`

File: src/test/java/com/fluentinterface/domain/TypedIDBuilder.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package com.fluentinterface.domain;
+
+import com.fluentinterface.builder.Builder;
+
+/**
```
Comment (davidmarquis) [inline src/test/java/com/fluentinterface/domain/TypedIDBuilder.java]: auto-generated comments, please remove

File: src/test/java/com/fluentinterface/domain/TypedID.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.fluentinterface.domain;
+
+/**
```
Comment (davidmarquis) [inline src/test/java/com/fluentinterface/domain/TypedID.java]: auto-generated comments, please remove

File: src/main/java/com/fluentinterface/utils/GenericsUtils.java [hunk 5]
```diff
@@ -18,7 +18,17 @@ private GenericsUtils() {}
             if (genericType instanceof ParameterizedType) {
                 ParameterizedType paramType = (ParameterizedType) genericType;
                 if (paramType.getRawType().equals(genericInterface)) {
-                    return (Class<?>) paramType.getActualTypeArguments()[0];
+
+                    Type type = paramType.getActualTypeArguments()[0];
+
+                    Class<?> rawType;
+                    if (type instanceof ParameterizedType) {
```
Comment (davidmarquis) [inline src/main/java/com/fluentinterface/utils/GenericsUtils.java]: please add a small comment (or refactor) to make the intent clearer

**Comment by eladlaufer** [comment]
Hi,

Don't want to nag, but did you happen to have the time to go over the latest commits?

Thanks

**Comment by davidmarquis** [comment]
OK I merged into master. I'll release a new version of the library that includes these changes soon but I probably won't have time today!

**Comment by eladlaufer** [comment]
Thanks :)

**Pull request merged**